WIP: Persist buffer state to history (for modes, tags and more)

Ambrevar opened the pull request:
# Description

This persists buffer modes (and more buffer data) in history.  This is useful so that on restart, modes are restored (which can be critical e.g. for privacy).

Fixes #2064 and #2583 (not yet).
Fixes #2036.

# How to test

- Make sure your cl-prevalence is up-to-date.

# Discussion

- [x] Is the design sane?
  In my opinion, yes, as it makes only minimal modifications to the Htree library, while adding a Nyxt core class that's easily extensible to add more buffer-related data to persist in the future.

- [x] Can we better serialize and deserialize URIs and timestamp?  The problem is that until now we used to serialize them as strings, but then how can Nyxt know which string is a URI, a timestamp, or just a string?
  An option would be to use reader macros, for instance `(local-time:enable-read-macros)`.  But that would break history backward compatibility.  Thoughts?

  - Another suggestion by @aartaka: Introduce `uri-sepcifier` and `timestamp-specifier` which accept strings, then write accessors to automatically replace the string with the structured version.  With a bit of MOP we could derive a general solution.
  - Better option: `coerce-slot`, which automatically coerce the value to the desired type.

# To do:

- [ ] Replace current dead buffers with serialized buffers.  Benefit: less risk for memory leaks.
- [x] Squash commits since this PR commit history is a mess.
- [x] Update cl-prevalence in submodules and Guix once patch is merged upstream. 
- [ ] Tests!
- [ ] Add a `tags` slot and persist it.
- [ ] Add tag-related commands.  See https://github.com/atlas-engineer/nyxt/issues/2583#issuecomment-1298379178
- [ ] Fix auto-rules which prevents any good use of this PR:
  - [ ] Currently auto-rules remove modes from the list of disabling them.
  - [ ] Reloading a page removes the manually enabled modes.  Example:
    - Go to https://en.wikipedia.org
    - Enable no-script-mode.
    - Reload: BOOM!  no-script-mode is gone.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [x] I have pulled from master before submitting this PR
- [x] There are no merge conflicts.
- [x] I've added the new dependencies as:
  - [x] ASDF dependencies,
  - [x] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [x] and Guix dependencies.
- [x] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [ ] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [ ] Documentation:
  - [x] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [x] I have updated the existing documentation to match my changes.
  - [x] I have commented my code in hard-to-understand areas.
  - [ ] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [ ] I have added a `migration.lisp` entry for all compatibility-breaking changes.
- [ ] Compilation and tests:
  - [ ] My changes generate no new warnings.
  - [ ] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [ ] New and existing unit tests pass locally with my changes.

aadcg commented:
@Ambrevar Could you please re-word the discussion section? It seems off.

aadcg commented:
Could you please explain the concept of a `tag` in this context?

aadcg commented on source/history.lisp:
```diff
@@ -45,6 +45,14 @@ It's a list of a form (Y &OPTIONAL X)."))
+    :documentation "List of modes that were last active for buffer."))
+  (:accessor-name-transformer (class*:make-name-transformer name))
+  (:documentation "Buffer state to restore when restoring history."))
+
```
`history.lisp` is loaded after `buffer.lisp`, so it's possible to access the `modes` slot from `modable-buffer`. Would it be a bad idea to dump the enabled modes to the history file before exiting Nyxt?

Wouldn't this functionality be achieved in a cleaner way via the new "`auto-mode`"?

Sorry if my understanding is completely off as I really don't know how we manage history!

Ambrevar replied:
History stores the list of buffer IDs and their histories.
When we restore the history, we recreate the corresponding buffers and associate them with these histories.

auto-rules associate URLs with modes.  Here we want to restore modes for specific buffers.

aadcg replied:
Clear.

aartaka replied:
Actually, we can try using history for all this `mast-active-modes` business of `auto-mode`/auto-rules (ʃ⌣́,⌣́ƪ)

aadcg commented on source/history.lisp:
```diff
@@ -344,7 +352,8 @@ Return non-NIL of history was restored, NIL otherwise."
                      (setf (gethash owner-id old-id->new-id) (id new-buffer))
-                     (setf (gethash (id new-buffer) new-owners) owner))))))
+                     (setf (gethash (id new-buffer) new-owners) owner)
+                     (enable-modes (modes (htree:data owner)) new-buffer))))))
```
I'm a bit confused by this. Could you explain?

aadcg replied:
To be more precise: `htree:data` can be whatever, so it feels a bit arbitrary to use it to restore the modes that were active in a buffer.

aartaka replied:
`htree:data` is an accessor for the arbitrarily-typed payload inside the owner/node. This is a design decision that we made to not tie ourselves to any particular design of histoy nodes and to make `htree` more reusable.

Ambrevar commented:
> @Ambrevar Could you please re-word the discussion section? It seems off.

Done.

> Could you please explain the concept of a `tag` in this context?

A buffer tag is an arbitrary string (or keyword?).  Users can then filter / categorize buffers.
See #2583 for a discussion.

Ambrevar commented:
Thought more about it: manually serializing modes and tags might be a bit shortsighted.

Instead, we should probably add a `s-serialization` method for buffers, which could work independently from history altogether.

For instance, create buffer B, serialize it with the new method, kill the buffer, then deserialize the serialized content to recreate the buffer.

aadcg commented:
That sounds better to me @Ambrevar!

Ambrevar commented:
By the way, this would address the long-standing issue of how to abstract "dead buffers".  Two (or three?) birds with one stone.

aartaka commented:
Yes, serializing buffers to history was the initial solution I thought of for buffer tagging :D

Ambrevar commented:
@aartaka It seems that auto-rules are broken.  For instance reloading a page disables the manually enabled modes, which is wrong.  (See the original post for an example).

Can you have a look?

Ambrevar commented:
Good news!  I'm able to serialize and restore buffers with all their modes!

aartaka commented:
> @aartaka It seems that auto-rules are broken. For instance reloading a page disables the manually enabled modes, which is wrong. (See the original post for an example).

Ahhhhhh, I might know the reason—I've removed the handler from `(en|dis)able-mode-hook`, and we lost the ability to preserve the most recent set of modes. I know how to fix it and will fix it tomorrow!

aartaka commented:
@Ambrevar, toggled modes preservation fixed in 16483000fbcc279ccfee050c71400bd52dddc692!

Ambrevar commented:
Thanks!

Ambrevar commented:
Before going further we need to merge #2654 and https://github.com/40ants/cl-prevalence/pull/22.

Ambrevar commented:
I've rebased and squashed. 

Prevalance still needs a fix: https://github.com/40ants/cl-prevalence/pull/25

Then there are a few more issues to iron out with this patch (like internal page URLs failing for some reason), but we are close!

aartaka commented:
Internal page URLs may fail because they are being loaded before the context has all the schemes initialized. But that's a wild guess :)

Ambrevar commented:
I'm working on the problem that URLs / timestamps are serialized as strings and thus need some manual code to be deserialized properly.  Of course the manual code is easy to write, the problem is that it needs to be set in many places which is cumbersome and error prone.

The following seems to work to ensure that `buffer`'s `url` slot is always a quri:uri:

```lisp
(defmethod (setf slot-value-using-class) :around (new-value
                                                  (class user-class)
                                                  (instance buffer)
                                                  slot)
  (if (eq 'url (c2mop:slot-definition-name slot))
      (call-next-method (ensure-url new-value) class instance slot)
      (call-next-method new-value class instance slot)))
```

Still, it needs to be done for all classes + it's verbose.  Can we do better?

I'm thinking of a few options:

1. Add a `:type-coercion` slot option that does the above by calling the value (a function) on the slot value and assigning the slot to the result.  Example

```lisp
(define-class ...
  ((foo "blah"
    :type string
    :type-coercion #'princ-to-string)
    ...))
```

Way less verbose, but still needs to be done for all slots.

2. Add a class option that does just the same for some types.
3. Add a global option / method that's called by user classes.  Then maybe this would do (untested):

```lisp
(defmethod (setf slot-value-using-class) :around (new-value
                                                  (class user-class)
                                                  instance
                                                  slot)
  (typecase (c2mop:slot-definition-type slot)
    ((quri:uri)
     (call-next-method (ensure-url new-value) class instance slot))
    (t
     (call-next-method new-value class instance slot))))
```

The drawback is that this is global, so if we want a bit more flexibility for a slot, we would need to override that method for a given class.  Or... just use a more flexible type.  So maybe that's not much of a drawback after all.

Thoughts?

aartaka commented:
> I'm working on the problem that URLs / timestamps are serialized as strings

But why not override cl-prevalence serialization for these?

Ambrevar commented:
> But why not override cl-prevalence serialization for these?

I guess you meant DEserialization.

We cannot do it based on the type, because the type is `string`, and we obviously cannot deserialize all strings to `quri:uri` or `local-time:timestamp`.

So we would have to do it for every quri:uri / timestamp slot of every class.  Which is what this pull request does.  It's cumbersome and error prone since it's easy to forget some slots.  Plus it's an extra burden for extension writers / user configuration.

Ambrevar commented:
Before this  can be tested properly #2895 must be addressed.

jmercouris commented:
Please reopen when ready.
